swarm/src/behaviour/listen_addresses.rs

```diff
@@ -0,0 +1,33 @@
+use crate::behaviour::{ExpiredListenAddr, FromSwarm, NewListenAddr};
+use crate::IntoConnectionHandler;
+use libp2p_core::Multiaddr;
+use std::collections::HashSet;
+
+/// Utility struct for tracking the addresses a [`Swarm`](crate::Swarm) is listening on.
+#[derive(Debug, Default, Clone)]
+pub struct ListenAddresses {
+    addresses: HashSet<Multiaddr>,
+}
+
+impl ListenAddresses {
+    /// Returns an [`Iterator`] over all listen addresses.
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = &Multiaddr> {
+        self.addresses.iter()
+    }
+
+    /// Feed a [`SwarmEvent`] to this struct.
+    pub fn on_event<THandler>(&mut self, event: &FromSwarm<THandler>)
```

Comment:
Should we here also:
```suggestion
    pub fn on_swarm_event<THandler>(&mut self, event: &FromSwarm<THandler>)
```
to match `NetworkBehaviour::on_swarm_event` ? (and the same for `ExternalAddresses`)